File: chart-pack.js
const VEDIC_ENGINE_URL = 'https://mocha-editor-monogamy.ngrok-free.dev/api/vedic-native';
const PACK_API = typeof API !== 'undefined' ? API : 'https://drishti-6n6f.onrender.com';

let packData = null;
let lastForm = null;

const DOMAIN_LABELS = {
  career: 'Career',
  wealth: 'Wealth',
  relationships: 'Relationships',
  health: 'Health',
  spirituality: 'Spirituality',
  travel: 'Travel',
};

const DOMAIN_ORDER = ['career', 'wealth', 'relationships', 'health', 'spirituality', 'travel'];

const DOMAIN_API_MAP = {
  career: 'career',
  wealth: 'wealth',
  relationship: 'relationships',
  relationships: 'relationships',
  health: 'health',
  spirituality: 'spirituality',
  travel: 'travel',
};

const BAND_LABELS = {
  high: 'High',
  moderate: 'Moderate',
  'low-moderate': 'Low–Moderate',
  low: 'Low',
  very_low: 'Very Low',
};

const YOGA_STRENGTH_RANK = { high: 0, medium: 1, low: 2, challenging: 3 };

const MD_THEMES = {
  Sun: 'The soul steps into its authority',
  Moon: 'The heart learns what it truly needs',
  Mars: 'Energy seeks its right direction',
  Mercury: 'The mind sharpens its true purpose',
  Jupiter: 'Wisdom expands into new territory',
  Venus: 'Beauty and harmony take centre stage',
  Saturn: 'The great teacher demands accountability',
  Rahu: 'Ambition meets the unknown',
  Ketu: 'The past releases what no longer serves',
};

function scoreToBandBar(score) {
  if (score >= 70) return { band: 'high', bar: 8 };
  if (score >= 45) return { band: 'moderate', bar: 5 };
  if (score >= 35) return { band: 'low-moderate', bar: 4 };
  if (score >= 25) return { band: 'low', bar: 3 };
  return { band: 'very_low', bar: 1 };
}

function parseDomainsFromContext(context) {
  const domains = {};
  if (!context) return domains;

  const lines = context.split('\n');
  const lineRe = /^\s+(CAREER|WEALTH|RELATIONSHIP|HEALTH|SPIRITUALITY|TRAVEL)\s+(\d+)\/100\s+\[(\w+)\]/;
  const triggerRe = /^\s+Triggers:\s*(.+)$/;

  lines.forEach((line, i) => {
    const m = line.match(lineRe);
    if (!m) return;

    const score = parseInt(m[2], 10);
    const { band, bar } = scoreToBandBar(score);
    const key = DOMAIN_API_MAP[m[1].toLowerCase()];

    let note = '';
    for (let j = i + 1; j < Math.min(i + 4, lines.length); j++) {
      const tm = lines[j].match(triggerRe);
      if (tm) {
        note = tm[1].trim();
        if (note.toLowerCase() === 'none') note = '';
        break;
      }
    }

    domains[key] = { band, bar, note, score };
  });

  return domains;
}

function yogaStrengthRank(yoga) {
  const eff = yoga.effective_strength || yoga.strength || 'medium';
  return YOGA_STRENGTH_RANK[eff] ?? 1;
}

function extractTopYogas(chart, limit = 3) {
  const yogas = chart.yogas || [];
  const ranked = [...yogas].sort((a, b) => yogaStrengthRank(a) - yogaStrengthRank(b));
  const top = ranked.slice(0, limit).map(y => ({
    name: y.name || 'Yoga',
    effect: y.desc || '',
    strength: y.effective_strength || y.strength || 'medium',
  }));
  return { top, total: yogas.length };
}

function formatBirthDetails(name, date, time, place) {
  try {
    const [y, m, d] = date.split('-').map(Number);
    const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    const dateFmt = `${String(d).padStart(2, '0')} ${months[m - 1]} ${y}`;
    return `${name} · ${dateFmt} · ${time} · ${place}`;
  } catch {
    return `${name} · ${date} · ${time} · ${place}`;
  }
}

function parseChartResponse(chart, form) {
  if (chart.error) throw new Error(chart.error);

  const asc = chart.ascendant || {};
  const planets = chart.planets || {};
  const dasha = chart.dasha || {};
  const rising = asc.sign || '—';
  const moon = (planets.Moon || {}).sign || '—';
  const sun = (planets.Sun || {}).sign || '—';
  const mahaLord = dasha.mahas || (dasha.current_maha || {}).lord || '—';

  const context = chart.karmi_context || chart.karmi_prompt_context || '';
  const domainScores = parseDomainsFromContext(context);
  DOMAIN_ORDER.forEach(key => {
    if (!domainScores[key]) {
      domainScores[key] = { band: 'moderate', bar: 5, note: 'Awaiting full domain analysis' };
    }
  });

  const { top: topYogas, total: totalYogaCount } = extractTopYogas(chart);

  return {
    identity: {
      name: form.name,
      birth_details: formatBirthDetails(form.name, form.date, form.time, form.place),
      rising,
      moon,
      sun,
    },
    theme: 'Theme coming soon',
    mahadasha: {
      lord: mahaLord,
      theme_line: MD_THEMES[mahaLord] || 'Your current planetary chapter unfolds.',
    },
    domain_scores: domainScores,
    top_yogas: topYogas,
    total_yoga_count: totalYogaCount,
  };
}

function prefillFromUrl() {
  const params = new URLSearchParams(window.location.search);
  const name = params.get('name');
  const date = params.get('date');
  const time = params.get('time');
  const place = params.get('place');

  if (name) document.getElementById('name').value = name;
  if (date) document.getElementById('dob').value = date;
  if (time) document.getElementById('tob').value = time;
  if (place) document.getElementById('place').value = place;
}

async function generateChartPack() {
  const name = document.getElementById('name')?.value?.trim();
  const dob = document.getElementById('dob')?.value;
  const tob = document.getElementById('tob')?.value;
  const place = document.getElementById('place')?.value?.trim();
  const btn = document.getElementById('generate-btn');

  if (!name || !dob || !tob || !place) {
    alert('Please fill in all birth details.');
    return;
  }
  if (!resolvedLat || !resolvedLon) {
    alert('Please select your birth city from the dropdown.');
    return;
  }

  lastForm = {
    name, date: dob, time: tob, place,
    lat: resolvedLat, lon: resolvedLon,
    timezone: resolvedTimezone,
  };

  const [year, month, day] = dob.split('-').map(Number);
  const [hour, minute] = tob.split(':').map(Number);
  const enginePayload = {
    year, month, day, hour, minute,
    lat: resolvedLat,
    lon: resolvedLon,
    utc_offset: resolvedTimezone,
  };

  const originalText = btn.textContent;
  btn.disabled = true;
  btn.textContent = 'Generating…';

  try {
    const res = await fetch(VEDIC_ENGINE_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'ngrok-skip-browser-warning': 'true',
      },
      body: JSON.stringify(enginePayload),
    });
    const chart = await res.json();
    if (!res.ok) throw new Error(chart.error || 'Chart computation failed');

    packData = parseChartResponse(chart, lastForm);
    showPackResults(packData);
  } catch (err) {
    alert(err.message || 'Could not generate chart pack. Please try again.');
  } finally {
    btn.disabled = false;
    btn.textContent = originalText;
  }
}

function showPackResults(data) {
  document.getElementById('pack-pre-result')?.classList.add('hidden');
  document.getElementById('pack-results')?.classList.remove('hidden');

  const ident = data.identity || {};
  document.getElementById('result-name').textContent = ident.name || '';
  document.getElementById('result-birth').textContent = ident.birth_details || '';
  document.getElementById('result-rising').textContent = ident.rising || '—';
  document.getElementById('result-moon').textContent = ident.moon || '—';
  document.getElementById('result-sun').textContent = ident.sun || '—';
  document.getElementById('result-theme').textContent = data.theme || '';

  const md = data.mahadasha || {};
  document.getElementById('result-md-line').textContent =
    `${md.lord || '—'} Mahadasha · ${md.theme_line || ''}`;

  renderDomainBars(data.domain_scores || {});
  renderYogas(data.top_yogas || [], data.total_yoga_count || 0);
  renderShareCard(data);
}

function renderDomainBars(scores) {
  const el = document.getElementById('domain-bars');
  if (!el) return;

  el.innerHTML = DOMAIN_ORDER.map(key => {
    const d = scores[key] || { bar: 5, band: 'moderate', note: '' };
    const bar = Math.min(10, Math.max(0, d.bar || 5));
    const cells = Array.from({ length: 10 }, (_, i) =>
      `<span class="pack-bar-cell${i < bar ? ' filled' : ''}"></span>`
    ).join('');
    const bandLabel = BAND_LABELS[d.band] || d.band;
    const note = d.note ? `<span class="pack-domain-note">${d.note}</span>` : '';
    return `
      <div class="pack-domain-row">
        <span class="pack-domain-label">${DOMAIN_LABELS[key]}</span>
        <div class="pack-bar-track">${cells}</div>
        <span class="pack-domain-band">${bandLabel}</span>
        ${note}
      </div>`;
  }).join('');
}

function renderYogas(yogas, total) {
  const list = document.getElementById('yoga-list');
  const more = document.getElementById('yoga-more');
  if (!list) return;

  list.innerHTML = yogas.map(y => `
    <div class="pack-yoga-item">
      <div class="pack-yoga-name">${y.name}</div>
      <div class="pack-yoga-effect">${y.effect}</div>
    </div>
  `).join('');

  const extra = (total || 0) - yogas.length;
  if (more) {
    more.textContent = extra > 0
      ? `+ ${extra} more yogas in your full chart`
      : '';
  }
}

function renderShareCard(data) {
  const ident = data.identity || {};
  const md = data.mahadasha || {};
  const line = `${ident.rising} Rising · ${ident.moon} Moon · ${md.lord} Mahadasha`;
  document.getElementById('share-line').textContent = line;
  document.getElementById('share-theme').textContent = data.theme || '';

  const shareText = `${line}\n${data.theme || ''}\n\nGet your chart pack: https://karmi.ai/pack`;
  const wa = document.getElementById('whatsapp-link');
  if (wa) {
    wa.href = `https://wa.me/?text=${encodeURIComponent(shareText)}`;
  }
  window._packShareText = shareText;
}

function copyShareText() {
  const text = window._packShareText || '';
  if (!text) return;
  navigator.clipboard.writeText(text).then(() => {
    alert('Copied to clipboard.');
  }).catch(() => {
    prompt('Copy this text:', text);
  });
}

function downloadFreePack() {
  if (!lastForm) {
    alert('Generate your chart pack first.');
    return;
  }
  const q = new URLSearchParams({
    name: lastForm.name,
    date: lastForm.date,
    time: lastForm.time,
    place: lastForm.place,
    lat: lastForm.lat,
    lon: lastForm.lon,
    timezone: lastForm.timezone,
  });
  window.location.href = `${PACK_API}/download-free-pack?${q.toString()}`;
}

document.addEventListener('DOMContentLoaded', prefillFromUrl);

File: chart_pack.py
"""Free chart pack — extract identity, domains, yogas from Karmi chart API."""

import re
from datetime import datetime

from theme_table import get_theme, get_md_theme

DOMAIN_KEYS = [
    "career", "wealth", "relationships",
    "health", "spirituality", "travel",
]

DOMAIN_API_MAP = {
    "career": "career",
    "wealth": "wealth",
    "relationship": "relationships",
    "relationships": "relationships",
    "health": "health",
    "spirituality": "spirituality",
    "travel": "travel",
}

STRENGTH_TO_FLOAT = {
    "high": 0.91,
    "medium": 0.82,
    "low": 0.70,
    "challenging": 0.65,
}


def score_to_band_bar(score: int):
    if score >= 70:
        return "high", 8
    if score >= 45:
        return "moderate", 5
    if score >= 35:
        return "low-moderate", 4
    if score >= 25:
        return "low", 3
    return "very_low", 1


def parse_domains_from_context(context: str) -> dict:
    """Parse domain lines from EventScorer karmi_context string."""
    domains: dict = {}
    if not context:
        return domains

    line_re = re.compile(
        r"^\s+(CAREER|WEALTH|RELATIONSHIP|HEALTH|SPIRITUALITY|TRAVEL)"
        r"\s+(\d+)/100\s+\[(\w+)\]",
        re.MULTILINE,
    )
    trigger_re = re.compile(
        r"^\s+Triggers:\s*(.+)$",
        re.MULTILINE,
    )

    lines = context.splitlines()
    for i, line in enumerate(lines):
        m = line_re.match(line)
        if not m:
            continue
        raw_name, score_str, band_raw = m.group(1), m.group(2), m.group(3)
        score = int(score_str)
        band, bar = score_to_band_bar(score)
        key = DOMAIN_API_MAP[raw_name.lower()]

        note = ""
        for j in range(i + 1, min(i + 4, len(lines))):
            tm = trigger_re.match(lines[j])
            if tm:
                note = tm.group(1).strip()
                if note.lower() == "none":
                    note = ""
                break

        domains[key] = {
            "band": band,
            "bar": bar,
            "note": note,
            "score": score,
        }

    return domains


def _yoga_strength(yoga: dict) -> float:
    eff = yoga.get("effective_strength") or yoga.get("strength") or "medium"
    return STRENGTH_TO_FLOAT.get(eff, 0.75)


def extract_top_yogas(chart: dict, limit: int = 3) -> tuple[list, int]:
    yogas = chart.get("yogas") or []
    ranked = sorted(yogas, key=_yoga_strength, reverse=True)
    top = []
    for y in ranked[:limit]:
        top.append({
            "name": y.get("name", "Yoga"),
            "effect": y.get("desc", ""),
            "strength": round(_yoga_strength(y), 2),
        })
    return top, len(yogas)


def _format_birth_details(name: str, date: str, time: str, place: str) -> str:
    try:
        dt = datetime.strptime(date, "%Y-%m-%d")
        date_fmt = dt.strftime("%d %b %Y")
    except ValueError:
        date_fmt = date
    return f"{name} · {date_fmt} · {time} · {place}"


def build_free_pack(chart: dict, name: str, date: str, time: str, place: str) -> dict:
    if "error" in chart:
        raise ValueError(chart["error"])

    asc = chart.get("ascendant") or {}
    planets = chart.get("planets") or {}
    dasha = chart.get("dasha") or {}
    rising = asc.get("sign", "—")
    moon = (planets.get("Moon") or {}).get("sign", "—")
    sun = (planets.get("Sun") or {}).get("sign", "—")

    maha_lord = dasha.get("mahas") or (dasha.get("current_maha") or {}).get("lord", "—")

    context = chart.get("karmi_context") or chart.get("karmi_prompt_context") or ""
    domain_scores = parse_domains_from_context(context)
    for key in DOMAIN_KEYS:
        if key not in domain_scores:
            domain_scores[key] = {
                "band": "moderate",
                "bar": 5,
                "note": "Awaiting full domain analysis",
            }

    top_yogas, total_yoga_count = extract_top_yogas(chart)
    theme = get_theme(rising, maha_lord)
    md_theme_line = get_md_theme(maha_lord)

    return {
        "identity": {
            "name": name,
            "birth_details": _format_birth_details(name, date, time, place),
            "rising": rising,
            "moon": moon,
            "sun": sun,
        },
        "theme": theme,
        "mahadasha": {
            "lord": maha_lord,
            "theme_line": md_theme_line,
        },
        "domain_scores": domain_scores,
        "top_yogas": top_yogas,
        "total_yoga_count": total_yoga_count,
    }


def build_free_md(pack: dict) -> str:
    ident = pack["identity"]
    md = pack["mahadasha"]
    domains = pack["domain_scores"]
    yogas = pack["top_yogas"]

    lines = [
        f"# {ident['name']} — Karmi Free Chart Pack",
        "",
        f"**Birth:** {ident['birth_details']}",
        "",
        "## Chart Identity",
        f"- **Rising:** {ident['rising']}",
        f"- **Moon:** {ident['moon']}",
        f"- **Sun:** {ident['sun']}",
        "",
        f"**Chart Theme:** *{pack['theme']}*",
        "",
        "## Life Right Now",
        f"**{md['lord']} Mahadasha** — {md['theme_line']}",
        "",
        "### Domain Scores",
    ]

    labels = {
        "career": "Career",
        "wealth": "Wealth",
        "relationships": "Relationships",
        "health": "Health",
        "spirituality": "Spirituality",
        "travel": "Travel",
    }
    for key in DOMAIN_KEYS:
        d = domains.get(key, {})
        bar = d.get("bar", 5)
        band = d.get("band", "moderate")
        note = d.get("note", "")
        bar_vis = "█" * bar + "░" * (10 - bar)
        lines.append(
            f"- **{labels[key]}** [{band}] {bar_vis} {note}"
        )

    lines += ["", "## Top Active Yogas"]
    for y in yogas:
        lines.append(f"- **{y['name']}** — {y['effect']}")

    extra = pack.get("total_yoga_count", 0) - len(yogas)
    if extra > 0:
        lines.append(f"\n*+ {extra} more yogas in your full chart*")

    lines += [
        "",
        "---",
        "",
        "Generated by [Karmi.ai](https://karmi.ai) · Upload this file to ChatGPT, Claude, or Gemini.",
        "",
        "Unlock your full chart pack at https://karmi.ai/pack",
    ]
    return "\n".join(lines)
